packages/framework/src/Model/Product/ProductFacade.php

```diff
@@ -194,6 +206,42 @@ public function edit(
         return $product;
     }
 
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param \Shopsys\FrameworkBundle\Model\Product\ProductData $productData
+     */
+    protected function refreshProductPromotion(Product $product, ProductData $productData): void
+    {
+        $promotionData = $productData->promotionXyData;
+
+        $promotionBuyQuantity = $promotionData?->buyQuantity;
+        $promotionFreeQuantity = $promotionData?->freeQuantity;
+        $currentPromotion = $product->getPromotionXy();
+
+        if ($promotionBuyQuantity === $currentPromotion?->getBuyQuantity() && $promotionFreeQuantity === $currentPromotion?->getFreeQuantity()) {
+            return;
+        }
+
+        if ($promotionBuyQuantity === null || $promotionFreeQuantity === null) {
+            if ($currentPromotion !== null) {
+                $product->setPromotionXy(null);
+                $this->em->remove($currentPromotion);
+            }
+
+            return;
+        }
+
+        $productPromotionXyData = $this->productPromotionXyDataFactory->create();
```

Comment:
This is probably the reason why promotion is always created as mentioned earlier